### PR TITLE
[add testcase] arc problems with recursive types

### LINKED ISSUE #9650: Destructors: problems with recursive types
Here is an example with recursive types that does work as expected if you replace `ptr` with `ref`, `alloc` with `new` and comment out `=destroy`, `=sink`, `=`.

```nim

import typetraits

type
  SharedPtr*[T] = object
    val: ptr tuple[atomicCounter: int, value: T]

  Node*[T] = object
    value: T
    next: SharedPtr[Node[T]]

  ForwardList*[T] = object
    first: SharedPtr[Node[T]]

proc `=destroy`*[T](p: var SharedPtr[T]) =
  if p.val != nil:
    let c = atomicDec(p.val[].atomicCounter)
    if c == 0:
      when not supportsCopyMem(T):
         `=destroy`(p.val[])
      dealloc(p.val)
    p.val = nil

proc `=`*[T](dest: var SharedPtr[T], src: SharedPtr[T]) =
  if dest.val != src.val:
    if dest.val != nil:
      `=destroy`(dest)
    if src.val != nil:
      discard atomicInc(src.val[].atomicCounter)
      dest.val = src.val

proc `=sink`*[T](dest: var SharedPtr[T], src: SharedPtr[T]) =
  if dest.val != nil and dest.val != src.val:
    `=destroy`(dest)
  dest.val = src.val

 
proc newSharedPtr*[T](val: sink T): SharedPtr[T] =
  result.val = cast[type(result.val)](alloc(sizeof(result.val[])))
  reset(result.val[])
  result.val.atomicCounter = 1
  result.val.value = val

proc isNil*[T](p: SharedPtr[T]): bool =
  p.val == nil

template `->`*[T](p: SharedPtr[T], name: untyped): untyped =
  p.val.value.name

proc createNode[T](val: T) : SharedPtr[ Node[T] ]=
  result = newSharedPtr(Node[T](value: val))

proc push_front*[T] (list: var ForwardList[T], val: T) =
  var newElem = createNode(val)
  newElem->next = list.first
  list.first = newElem

proc pop_front*[T] (list: var ForwardList[T]) =
  let head = list.first
  list.first = head->next

proc `$`*[T] (list: ForwardList[T]): string =
  result = "["
  var head = list.first
  while not head.isNil:
    result &= $(head->value) & ", "
    head = head->next
  result &= ']'

var x: ForwardList[int]
x.push_front(1)
x.push_front(2)
x.push_front(3)

echo x

x.pop_front()
x.pop_front()
echo x

x.pop_front()
echo x
```
Example currently fails with ``Error: cannot evaluate 'sizeof' because its type is not defined completely``, it is weird as it has no problems computing sizeof when refs are used.

You can replace sizeof with some constant like 128 to workaround it.

### PR DESCRIPTION
**Pull request by ringabout**
close #9650

### CODE REVIEW

**Comment by Araq** [comment]
CI failure seems related, please take a look.

**Comment by ringabout** [comment]
ok

File: tests/arc/t9650.nim [hunk 1]
```diff
@@ -0,0 +1,87 @@
+discard """
+  matrix: "--gc:arc --expandArc:toString"
```
Comment (ringabout) [inline tests/arc/t9650.nim]:
```suggestion
  matrix: "--gc:arc"
```

**Comment by ringabout** [comment]
CI failure seems related with compile option. I will investigate it later.

**Pull request merged**